File: 213.cpp
class Solution {
public:

     int f(int i, vector<int>& nums, vector<int>& dp){
        if(i < 0)return 0;
        if(dp[i] != -1)return dp[i];
        return dp[i] = max(nums[i]+f(i-2, nums, dp), max(f(i-2, nums, dp), f(i-1, nums, dp)) );
    }

    int rob(vector<int>& nums) {
        if(nums.size() == 1)return nums[0];
        vector<int> dp1(nums.size(), -1);
        int res1 =  f(nums.size()-2, nums, dp1);
        reverse(nums.begin(), nums.end());
        nums.pop_back();
        vector<int> dp2(nums.size(), -1);
        int res2 = f(nums.size()-1, nums, dp2);
        return max(res1, res2);
    }
};

File: Python/213.py
class Solution:
    def f(self, ind, nums, dp):
        if ind < 0:
            return 0
        if dp[ind] != -1:
            return dp[ind]
        # take
        take = nums[ind] + self.f(ind - 2, nums, dp)
        not_take = self.f(ind-1, nums, dp)
        dp[ind] = max(take, not_take)
        return dp[ind]

    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        a1 = nums[1:]
        a2 = nums[:-1]
        dp1 = [-1]*len(a1)
        dp2 = [-1]*len(a1)
        return max(self.f(len(a1)-1, a1, dp1), self.f(len(a2)-1, a2, dp2))
